compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIETransferStridedAccessPattern.cpp

```diff
@@ -0,0 +1,399 @@
+/// empty or 1-d vectors.
+static bool isL2AddressingLinear(SmallVector<OpFoldResult> &dmaOffsets,
+                                 SmallVector<OpFoldResult> &dmaSizes,
+                                 SmallVector<OpFoldResult> &dmaStrides) {
+  assert(dmaOffsets.size() == dmaSizes.size() &&
+         dmaOffsets.size() == dmaStrides.size() &&
+         "expected same number of source offsets and sizes");
+  if (dmaOffsets.size() == 0) return true;
+  if (dmaOffsets.size() != 1) return false;
+  if (!isConstantIntValue(dmaOffsets[0], 0)) return false;
+  if (!isConstantIntValue(dmaStrides[0], 1)) return false;
+  return true;
+}
+
+/// Utility to check if all users of the connection op statisfy the conditions
+/// for dma access pattern transfer.
+static FailureOr<bool> checkConnectionUsers(AMDAIE::ConnectionOp connectionOp) {
+  for (Operation *user : connectionOp->getUsers()) {
+    // Check if L3 addressing is combinable.
+    if (auto dmaOp = dyn_cast<AMDAIE::NpuDmaCpyNdOp>(user)) {
+      if (dmaOp.hasSourceAddressing() && dmaOp.hasTargetAddressing()) {
+        dmaOp.emitOpError()
+            << "should not have both source and target addressing";
+        return failure();
+      }
+      if (!dmaOp.hasSourceAddressing() && !dmaOp.hasTargetAddressing()) {
+        dmaOp.emitOpError() << "should have either source or target addressing";
+        return failure();
+      }
+
+      SmallVector<OpFoldResult> dmaOffsets;
+      SmallVector<OpFoldResult> dmaSizes;
+      SmallVector<OpFoldResult> dmaStrides;
+      if (dmaOp.hasSourceAddressing()) {
+        dmaOffsets = dmaOp.getSourceMixedOffsets();
+        dmaSizes = dmaOp.getSourceMixedSizes();
+        dmaStrides = dmaOp.getSourceMixedStrides();
+      }
+      if (dmaOp.hasTargetAddressing()) {
+        dmaOffsets = dmaOp.getTargetMixedOffsets();
+        dmaSizes = dmaOp.getTargetMixedSizes();
+        dmaStrides = dmaOp.getTargetMixedStrides();
+      }
+      size_t dimForCombine = dmaSizes.size();
+      if (!isL3AddressingCombinable(dmaOffsets, dmaSizes, dmaStrides,
+                                    dimForCombine)) {
+        return false;
+      }
+    }
+    // Check if L2 addressing is linear.
+    if (auto circularDma = dyn_cast<AMDAIE::NpuCircularDmaCpyNdOp>(user)) {
+      if (circularDma.hasSourceAddressing() &&
+          circularDma.hasTargetAddressing()) {
+        circularDma.emitOpError()
+            << "should not have both source and target addressing";
+        return failure();
+      }
+
+      SmallVector<OpFoldResult> circularOffsets;
+      SmallVector<OpFoldResult> circularSizes;
+      SmallVector<OpFoldResult> circularStrides;
+      if (circularDma.hasSourceAddressing()) {
+        circularOffsets = circularDma.getSourceMixedOffsets();
+        circularSizes = circularDma.getSourceMixedSizes();
+        circularStrides = circularDma.getSourceMixedStrides();
+      }
+      if (circularDma.hasTargetAddressing()) {
```

Comment:
Npu circular dma op could have both source and target addressing empty, so I'm not using "else" here.

Reply:
Correct. But since you're ensuring there's going to be either of those two in the preceding checks we could've just used `else` here.

I'm marking this resolved anyway pertaining to Jorn's comment about basing the check on MemSpace Attr.